api/plugin.go

```diff
@@ -0,0 +1,474 @@
+			p, err := c.Holder.PluginHolder.Get(conf.ModulePath)
+			name := "[UNNAMED]"
+			if err == nil {
+				if info := p.GetPluginInfo(); info.Name != "" {
+					name = info.Name
+				} else {
+					name = info.ModulePath
+				}
+			}
+			result = append(result, model.PluginConfExternal{
+				ID:           conf.ID,
+				Name:         name,
+				Token:        conf.Token,
+				ModulePath:   conf.ModulePath,
+				Enabled:      conf.Enabled,
+				Capabilities: c.getPluginCapabilities(inst),
+			})
+		}
+	}
+	ctx.JSON(200, result)
+}
+
+func (c *PluginAPI) getPluginCapabilities(inst wrapper.PluginInstance) []string {
+	capabilities := make([]string, 0)
+	if inst.GetConfigurer() != nil {
+		capabilities = append(capabilities, "configurer")
+	}
+	if inst.GetDisplayer() != nil {
+		capabilities = append(capabilities, "displayer")
+	}
+	if inst.GetMessenger() != nil {
+		capabilities = append(capabilities, "messenger")
+	}
+	if inst.GetStorager() != nil {
+		capabilities = append(capabilities, "storager")
+	}
+	if inst.GetWebhooker() != nil {
+		capabilities = append(capabilities, "webhooker")
+	}
+	return capabilities
+}
+
+// EnablePlugin enables a plugin.
+// swagger:operation POST /plugin/:id/enable plugin enablePlugin
+//
+// Enable a plugin.
+//
+// ---
+// consumes: [application/json]
+// produces: [application/json]
+// security: [clientTokenHeader: [], clientTokenQuery: [], basicAuth: []]
+// responses:
+//   200:
+//     description: Ok
+//   401:
+//     description: Unauthorized
+//     schema:
+//         $ref: "#/definitions/Error"
+//   500:
+//     description: Internal Server Error
+//     schema:
+//         $ref: "#/definitions/Error"
+func (c *PluginAPI) EnablePlugin(ctx *gin.Context) {
+	withID(ctx, "id", func(id uint) {
+		if ins := c.DB.GetPluginConfByID(id); ins == nil {
+			ctx.AbortWithError(401, errors.New("unknown plugin token"))
+			return
+		}
+		err := c.setPluginEnableStatus(id, true)
+		if err != nil {
+			ctx.AbortWithError(500, err)
+		}
+	})
+}
+
+// DisablePlugin disables a plugin.
+// swagger:operation POST /plugin/:id/disable plugin disablePlugin
+//
+// Disable a plugin.
+//
+// ---
+// consumes: [application/json]
+// produces: [application/json]
+// security: [clientTokenHeader: [], clientTokenQuery: [], basicAuth: []]
+// responses:
+//   200:
+//     description: Ok
+//   401:
+//     description: Unauthorized
+//     schema:
+//         $ref: "#/definitions/Error"
+//   500:
+//     description: Internal Server Error
+//     schema:
+//         $ref: "#/definitions/Error"
+func (c *PluginAPI) DisablePlugin(ctx *gin.Context) {
+	withID(ctx, "id", func(id uint) {
+		if ins := c.DB.GetPluginConfByID(id); ins == nil {
+			ctx.AbortWithError(401, errors.New("unknown plugin token"))
+			return
+		}
+		err := c.setPluginEnableStatus(id, false)
+		if err != nil {
+			ctx.AbortWithError(500, err)
+		}
+	})
+}
+
+func (c *PluginAPI) setPluginEnableStatus(id uint, enabled bool) error {
+	c.mutex.Lock()
```

Comment:
why do we need a mutex here?

Reply:
To prevent disable and enable being called simultaneously creating an inconsistency with the enabled state of the plugin instance and the record in the database